From 8265120c73868152dc3a8c4e6a97536dd5cf6492 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Thu, 4 May 2023 10:40:02 +0800
Subject: [PATCH 1623/1646] [Huawei] block: support to disable debugfs by
 default

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188711

----------------------------------------

All block layer debugfs is enabled by default, and it can be disabled by
adding "blk_mq.enable_debugfs=0" to boot cmd. Noted that user can still
enable or disable debugfs by sysfs for each disk.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig |  1 +
 arch/sw_64/configs/openeuler_defconfig |  1 +
 arch/x86/configs/openeuler_defconfig   |  1 +
 block/Kconfig                          | 11 +++++++++++
 block/blk-mq.c                         | 10 ++++++++++
 5 files changed, 24 insertions(+)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 3a9d59fb8326..865a5f9050d1 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -917,6 +917,7 @@ CONFIG_BLK_CGROUP_IOCOST=y
 CONFIG_BLK_CGROUP_LEGACY_IOCOST=y
 CONFIG_BLK_WBT_MQ=y
 CONFIG_BLK_DEBUG_FS=y
+CONFIG_BLK_DEBUG_FS_SWITCH=y
 CONFIG_BLK_DEBUG_FS_ZONED=y
 # CONFIG_BLK_SED_OPAL is not set
 # CONFIG_BLK_INLINE_ENCRYPTION is not set
diff --git a/arch/sw_64/configs/openeuler_defconfig b/arch/sw_64/configs/openeuler_defconfig
index 0e77721dae36..792d33743d59 100644
--- a/arch/sw_64/configs/openeuler_defconfig
+++ b/arch/sw_64/configs/openeuler_defconfig
@@ -532,6 +532,7 @@ CONFIG_BLK_DEV_INTEGRITY_T10=y
 # CONFIG_BLK_CGROUP_IOLATENCY is not set
 # CONFIG_BLK_CGROUP_IOCOST is not set
 CONFIG_BLK_DEBUG_FS=y
+CONFIG_BLK_DEBUG_FS_SWITCH=y
 # CONFIG_BLK_SED_OPAL is not set
 # CONFIG_BLK_INLINE_ENCRYPTION is not set
 
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index b1ed7f628a7d..ede44699371c 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -908,6 +908,7 @@ CONFIG_BLK_CGROUP_IOCOST=y
 CONFIG_BLK_CGROUP_LEGACY_IOCOST=y
 CONFIG_BLK_WBT_MQ=y
 CONFIG_BLK_DEBUG_FS=y
+CONFIG_BLK_DEBUG_FS_SWITCH=y
 CONFIG_BLK_DEBUG_FS_ZONED=y
 # CONFIG_BLK_SED_OPAL is not set
 # CONFIG_BLK_INLINE_ENCRYPTION is not set
diff --git a/block/Kconfig b/block/Kconfig
index eb1244a83ef2..e3d4619d956a 100644
--- a/block/Kconfig
+++ b/block/Kconfig
@@ -193,6 +193,17 @@ config BLK_DEBUG_FS
 	Unless you are building a kernel for a tiny system, you should
 	say Y here.
 
+config BLK_DEBUG_FS_SWITCH
+	bool "Disable/Enable block layer debugfs by demands"
+	default y
+	depends on BLK_DEBUG_FS
+	help
+	All block layer debugfs is enabled by default, enabling this option
+	allow user to disable all debugfs by module parameters. Noted that
+	user can still enable or disable debugfs by sysfs for each disk.
+
+	If memory resources is limited, you should say Y here.
+
 config BLK_DEBUG_FS_ZONED
        bool
        default BLK_DEBUG_FS && BLK_DEV_ZONED
diff --git a/block/blk-mq.c b/block/blk-mq.c
index c7610a752ba8..c3f39cc9c2a7 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -44,6 +44,11 @@
 bool mq_unfair_dtag = true;
 module_param_named(unfair_dtag, mq_unfair_dtag, bool, 0444);
 
+#ifdef CONFIG_BLK_DEBUG_FS_SWITCH
+bool enable_debugfs = true;
+module_param_named(enable_debugfs, enable_debugfs, bool, 0444);
+#endif
+
 static DEFINE_PER_CPU(struct list_head, blk_cpu_done);
 
 static void blk_mq_poll_stats_start(struct request_queue *q);
@@ -3495,6 +3500,11 @@ struct request_queue *blk_mq_init_allocated_queue(struct blk_mq_tag_set *set,
 	q->tag_set = set;
 
 	q->queue_flags |= QUEUE_FLAG_MQ_DEFAULT;
+
+#ifdef CONFIG_BLK_DEBUG_FS_SWITCH
+	if (!enable_debugfs)
+		blk_queue_flag_clear(QUEUE_FLAG_DEBUGFS, q);
+#endif
 	if (set->nr_maps > HCTX_TYPE_POLL &&
 	    set->map[HCTX_TYPE_POLL].nr_queues)
 		blk_queue_flag_set(QUEUE_FLAG_POLL, q);
-- 
2.17.1

